### ISSUE TITLE
whenever with a Seq blocks other whenevers

### ISSUE THREAD

**Issue by b2gills** [rakudo/rakudo#1612]
## The Problem

`whenever ｢Seq｣ {…}` and `whenever ｢Seq｣.Supply {…}` blocks other `whenever`s until it has produced at least one value.

(This makes it difficult to use the `react`/`supply` + `whenever` construct with `$*IN.lines`)

## Example Code

``` perl6
#! /usr/bin/env perl6
use v6.d.PREVIEW;
#use v6.c;

say $*PERL.compiler;

my $seq = gather {
  sleep 2;   # delay the first value
  for ^3 {
    .take;
    sleep ⅔; # delay between values
  }
}

# this is to show that it is blocking the other whenever
Promise.in(1+⅞).then: { say 'Should have printed several lines by now' }

react {
  whenever Supply.interval(¼) {
    say "$_ Supply";
  }
  whenever $seq.Supply {
    LAST done;
    say "$_ Seq";
  }
}
```

## Expected Result

```
rakudo (2018.02.1.146.g.52.e.66.ad)
0 Supply
1 Supply
2 Supply
3 Supply
4 Supply
5 Supply
6 Supply
7 Supply
Should have printed several lines by now
0 Seq
8 Supply
9 Supply
10 Supply
1 Seq
11 Supply
12 Supply
13 Supply
2 Seq
14 Supply
15 Supply
3 Seq
16 Supply
17 Supply
18 Supply
4 Seq
19 Supply
20 Supply
21 Supply
```

## Actual Result

```
rakudo (2018.02.1.146.g.52.e.66.ad)
Should have printed several lines by now
0 Supply
1 Supply
2 Supply
3 Supply
4 Supply
5 Supply
6 Supply
7 Supply
8 Supply
0 Seq
9 Supply
10 Supply
1 Seq
11 Supply
12 Supply
13 Supply
2 Seq
14 Supply
15 Supply
16 Supply
3 Seq
17 Supply
18 Supply
4 Seq
19 Supply
20 Supply
21 Supply
```

## Workaround

``` perl6
# This is suboptimal as it will continue to iterate the Seq
# until it is fully consumed
sub Supply-from-Seq ( Seq:D $sequence --> Supply:D ) {
  my $supplier = Supplier.new;
  start {
    $sequence.map: { $supplier.emit($_) }
    $supplier.done();
  }
  $supplier.Supply;
}

…

whenever Supply-from-Seq($seq) {
  LAST done;
  say "$_ Seq";
}

```

(This is the code used to generate the expected result above)

## Possible Fix

Make the supply from `Seq.Supply` not block on the first value being available, and have `whenever ｢Seq｣ {…}` act as if you had called it with `whenever ｢Seq｣.Supply {…}`.

Note that `Seq.Supply` uses the generic method candidate in `Any`.

## Possible ROAST Test

``` perl6
todo '｢whenever｣ on a Seq blocks other ｢whenever｣s';
{
  my $timeout = 1/32;

  # Use a Promise so that it can't be changed once set.
  # Protects against other possible 'react'/'whenever' bugs
  # from causing a false pass result.
  my $pass = Promise.new;

  my Seq $seq = gather {
    # must use a blocking construct here
    sleep $timeout;
    take Nil;
  }

  react {
    whenever $seq.Supply {
      # Should never get here if $timeout is large enough
      $pass.keep( False );
      done;
    }
    whenever Nil {
      $pass.keep( True );
      done;
    }
  }

  ok $pass.result, "whenever with a delayed Seq doesn't block other whenevers"
}
```

I can add this test myself if the semantics are acceptable.

It might be advisable to make sure the `supply` + `whenever` construct also creates a Supply that doesn't cause the same sort of blocking behaviour.

**Comment by jnthn** [comment]
There's two sets of semantics to consider here: those of `react`/`whenever`, and those of the `Supply` coercer on a `Seq`.

So far as the first goes, the current semantics are as intended, and in line with the goal of the `react`/`supply`/`whenever` construct, which is to tame concurrency rather than to introduce it. Its one-message-at-a-time semantics extend to the setup phase (which is seen as a kind of zero message), in order that one may be confident all the setup logic in the `supply`/`react` block will run to completion before any further messages are processed. The expectation of this example is that it will somehow magically start a new thread that `sleep` can block. That isn't a reasonable expectation.

There has been significant work to make sure that when doing a `tap` of a `Supply` that wishes to work "synchronously", it can do so, but it must be cooperative. For example, switching from `sleep` to `await Promise.in(...)` achieves that:

```
#! /usr/bin/env perl6
use v6.d.PREVIEW;
#use v6.c;

say $*PERL.compiler;

my $seq = gather {
  await Promise.in(2);   # delay the first value
  for ^3 {
    .take;
    await Promise.in(⅔); # delay between values
  }
}

# this is to show that it is blocking the other whenever
Promise.in(1+⅞).then: { say 'Should have printed several lines by now' }

react {
  whenever Supply.interval(¼) {
    say "$_ Supply";
  }
  whenever $seq.Supply {
    LAST done;
    say "$_ Seq";
  }
}
```

And does indeed produce the desired output. Of course, that's not a general solution for this problem. The general solution is to explicitly say that we want the `Seq` evaluation to run as a separate task in the thread pool, which can be achieved like this:

```
#! /usr/bin/env perl6
use v6.d.PREVIEW;
#use v6.c;

say $*PERL.compiler;

my $seq = gather {
  sleep 2;   # delay the first value
  for ^3 {
    .take;
    sleep ⅔; # delay between values
  }
}

# this is to show that it is blocking the other whenever
Promise.in(1+⅞).then: { say 'Should have printed several lines by now' }

react {
  whenever Supply.interval(¼) {
    say "$_ Supply";
  }
  whenever Supply.from-list($seq, scheduler => $*SCHEDULER) {
    LAST done;
    say "$_ Seq";
  } 
} 
```

Which brings us to the second set of semantics, which is those of the `Seq.Supply` method. Arguably, `Seq.Supply` could default to scheduling itself using `$*SCHEDULER`. Note, however, that doing so would mean that the program `(1, 2 ... 10).Supply.tap(&say)` no longer produces any output, but instead terminates because the events all occur on another thread. So this would be a decidedly visible change, even if arguably useful.

**Comment by b2gills** [comment]
@jnthn That was just example code, the real problem is when you try and use `$*IN.lines` which blocks **everything** until there has been at least one line seen, after that it doesn't block in the same way. I have dealt with 2 or 3 cases of this on Stack Overflow. Also the first workaround I came up with doesn't seem to work anymore.

That is everything is otherwise fine except for the very first call of `.pull-one`.

I think it is because `Seq.Supply` uses the generic one that just calls `.list.Supply`

It would be nice if `$*IN.lines.Supply` reads lines as they are consumed rather than as soon as they are seen. I would think that `Supply.from-list($*IN.lines, scheduler => $*SCHEDULER)` does the latter.

I mean `$*IN.lines.Supply.throttle(1,¼)` probably shouldn't read as fast as possible.

**Comment by jnthn** [comment]
> It would be nice if $*IN.lines.Supply reads lines as they are consumed rather than as soon as they are seen. I would think that Supply.from-list($*IN.lines, scheduler => $*SCHEDULER) does the latter.

No, as of the big `Supply` concurrency refactor last year, it will read them as they are consumed. That's because the thread reading the lines then does the `emit`, and thus runs the `whenever`, and any downstream `whenever`s (and if it can't run the `whenever` right now, it soft-blocks until its turn). Thus it pays the cost of the processing of that line before it reads another.

(It used to be that the `supply` and `react` constructs piled up messages to process, unlike every other construct which enforced "sender pays". Now `supply` and `react` also enforce sender pays, except in the case of recursion (sending a message to yourself), where it resorts to the thread pool to break the deadlock. That case aside, there's now backpressure on fast emitters by virtue of them paying the processing costs.)

**Comment by b2gills** [comment]
I still think it is a bug that the first call of `Iterator.pull-one` seems to be blocking the the `react` initialization. Talk about action at a distance.

The following provides the correct behaviour if you swap out the `$seq.Supply` `whenever` block for the `whenever` seen below.  
(Or rather it seems as though it works right even though it is still blocking on the first `Iterator.pull-one`; it just doesn't have to wait as long.)

``` perl6
my \DON'T-BLOCK = Any.new; # take doesn't work with Mu.new

my $wrapped = gather {
  # add initial value to be thrown away
  take DON'T-BLOCK;

  .take for @$seq;
}
```
``` perl6
  …
  whenever $wrapped.Supply {
    # throw away initial value
    # next if $_.WHICH eqv  DON'T-BLOCK.WHICH;
    once next;

    LAST done;
    say "$_ Seq";
  }
  …
```

**Comment by lizmat** [comment]
Is this still an issue?
